sys/net/fib: added function to request a set of destination addresses

BytesGalore opened the pull request:
This PR introduces an access function to the FIB, that collects all destination addresses matching specific prefix.
(This is meant as a base for discussion)

Rationale:
Using the FIB with RPL requires in storing MOP that all destination addresses reachable from a node, which are constructed by receiving DIOs and all target options from all received DAOs recursively, are put in the DAO messages disseminated by the node.
The former FIB did not provided access for a set of destination addresses matching a specific prefix.

cgundogan commented on sys/net/network_layer/fib/fib.c:
```diff
@@ -421,6 +421,41 @@ int fib_get_next_hop(kernel_pid_t *iface_id,
+    for (size_t i = 0; i < FIB_MAX_FIB_TABLE_ENTRIES; ++i) {
+        if( (fib_table[i].global_flags & bitmap) == (type & bitmap) ) {
+
+            if( dst_set != NULL && found_entries < (*dst_set_size)-1 ) {
```
Could you move the check `dst_set != NULL` outside of the loop?

BytesGalore replied:
Basically yes, but the it would require a construct like:
``` C
if( dst_set != NULL ){
 for(...) ... // like above without the check
} else {
 for(...) ... // just counting the entries
}
```

cgundogan replied:
I see. Actually I thought that this function should return an error when no `dst_set` was specified. Anyhow, can you extend the doc that this function still counts and returns all matching entries even when they are not copied to `dst_set`?

And add parens to the condition, please (:

BytesGalore replied:
Added parens and wrote a line that the matching count is being determined.

The idea to allow a `dst_set == NULL` and continue, is that if you're interested at first in the current number of matching entries, to check if your out buffer can hold them, you probably don't want to have any copy operations.

cgundogan commented on sys/include/net/ng_fib.h:
```diff
@@ -122,6 +130,27 @@ int fib_get_next_hop(kernel_pid_t *iface_id,
 /**
+* @brief provides a set of destination addresses matching the bits specified with
+* the provided bitmap and type.
+* If the out buffer is insufficient low, the function will continue to count the
```
could you also add the fact, that the buffer is allowed to be null? Something like:
>If the out buffer is insufficient low or null, ...

BytesGalore replied:
sure, sorry for the delay.

cgundogan commented:
besides my comment this PR looks good and needs squashing

cgundogan commented:
does anyone else wants to participate in this discussion (because it's marked as RFC)? Otherwise, my ACK holds and my intention is to merge this ASAP

BytesGalore commented:
squashed and rebased.

cgundogan commented:
I guess I am satisfied with this PR and I am in favor of dropping the RFC label here and opt for a quick merge, since we need this functionality for #2765 . What do you say @BytesGalore @OlegHahm ?

BytesGalore commented:
@cgundogan sure, removed the RFC label

BytesGalore commented:
in a followup PR we could make this feature optional/exclusive when RPL is loaded

emmanuelsearch commented:
Do we need to think about a rationale such that their are no flag collisions between different protocols that write into the FIB? Like a prefix based on protocol ID, and then the rest of the bits are managed per protocol?

BytesGalore commented:
@emmanuelsearch definitely, I think an `enum` with define protected members will do, similar to [1]

[1]  https://github.com/RIOT-OS/RIOT/blob/master/sys/include/net/ng_nettype.h#L89

BytesGalore commented:
I don't expect that we will have like 100 routing-protocols available and using more than 3 at the same time on a node

cgundogan commented:
@BytesGalore are `global_flags` and `next_hop_flags` only used for routing protocol distinction? What other uses for them are out there? Introducing a mechanism for collision-free type definitions mentioned by @emmanuelsearch is favorable. Personally, I have no strong opinion of it being an `enum` or somehow encoding this information by using specific prefixes. I would argue however that this should be done in a separate PR and most preferably prior to this one

BytesGalore commented:
@cgundogan the `_flags` are used also to identify the address type, e.g. `NG_NETTYPE_IPV6`. 
Its beneficial to know what type of address hides behind the address bytes, and is necessary when we provide using multiple Protocols with different address types.

But I think you're both are right. A collision free mechanism based on protocol IDs and additional management bits is the right way to do it and this should be a done/merged prior to this PR.

cgundogan commented:
@BytesGalore do you have a concept in mind for such a mechanism?

BytesGalore commented:
@cgundogan after rethinking a bit I think using flags/protocol IDs for that purpose is not necessary. 
A routing protocol is responsible for a specific prefix. All destination addresses are indirectly "assigned" to a protocol by the prefix. 
So asking the FIB for a set of destination addresses based on a given prefix seems more reasonable for me here.

BytesGalore commented:
The prototype would change to:
``` C
int fib_get_destination_set(uint32_t* prefix, size_t prefix_length, fib_destination_set_entry_t *dst_set, size_t* dst_set_size);
```
@cgundogan what do you think, any cons on change it this way?

cgundogan commented:
@BytesGalore I guess your suggestion is reasonable. Assigning the routing protocols to handle certain prefixes may be the right way. To stress a point: What happens if prefix ranges overlap? longest prefix matches?

BytesGalore commented:
@cgundogan overlapping should MUST NOT happen since then both (or multiple) routing protocols would take responsibility/royalty over the intersected address range.
This way the overlapping entries would be repeatedly overwritten form multiple sources with distinct values, preventing a healthy topology for each routing protocol using that shared range.
In consequence such approach will eventually result in routing loops since no routing protocol knows the current state of the topology in the overlapping range.

BytesGalore commented:
ok, changed the approach to use a prefix match instead of flags, and rebased.

cgundogan commented on sys/net/network_layer/fib/universal_address.c:
```diff
@@ -197,6 +197,35 @@ int universal_address_compare(universal_address_container_t *entry,
 }
 
+int universal_address_compare_prefix(universal_address_container_t *entry,
+                              uint8_t *prefix, size_t prefix_size)
```
the `prefix_size` refers to the full length of the address type used in `prefix`, right?
In case of a prefix for ipv6 addresses `2001:db8:3c4d::/48`, I would  still use sizeof(ng_ipv6_addr) instead of 48? Therefore, a "prefix address" MUST have trailing zeros? Can you add to the doc, that trailing zeros are a MUST for prefix checks, which are not ::/128 (in case of ipv6)?

cgundogan replied:
FYI:
https://github.com/RIOT-OS/RIOT/blob/master/sys/net/network_layer/ng_ipv6/addr/ng_ipv6_addr.c#L22

BytesGalore commented:
@cgundogan I adjusted the doxy and put some words about the full address size requirement. 
And I changed the prefix match approach to take bits into account (thx for the hint :) ).

This changes the API slightly for `universal_address_compare()` to have an additional return value and requires the provided address_size in bits than in bytes. 
Though I don't know if this must be tagged since this function is only used by the FIB internally (by now).

cgundogan commented:
@BytesGalore I like it now (: Let's rebase, squash and merge it

BytesGalore commented on sys/include/net/ng_fib.h:
```diff
@@ -34,6 +34,14 @@ typedef struct rrp_address_msg_t {
+ * @brief entry used to collect available destinations
+ */
+typedef struct fib_destination_set_entry_t {
+    uint8_t dest[16];    /**< The destination address */
```
we could wait for #3022 to set the size of `dest[]` to `UNIVERSAL_ADDRESS_SIZE` instead of this magic number.
I can also make  a PR to change this later on, since RPL is using ipv6 addresses of 16 bytes anyways

cgundogan replied:
@BytesGalore I'm in favor of getting this PR merged asap. I think the fixes to your address_compare functions are significant. Magic numbers can be replaced after #3022

BytesGalore commented:
@cgundogan ok, rebased and squashed

cgundogan commented:
I think there is a bug somewhere in the compare function. When I update the lifetime of the default route `::`, it tells me

cgundogan commented:
I get the same behaviour on master

cgundogan commented:
It may be the part where you search for the first trailing zero.. What happens if everything is `0`, like it's the case for `::`?

BytesGalore commented:
@cgundogan I know this issue, the fix for this is on the way #2783 :)

cgundogan commented:
@BytesGalore ah nice! I completely overlooked this PR. Will test it.

cgundogan commented:
@BytesGalore I tested this and can confirm that it works. ACK

OlegHahm commented on sys/include/net/ng_fib.h:
```diff
@@ -37,6 +37,14 @@ typedef struct rp_address_msg_t {
+ * @brief entry used to collect available destinations
+ */
+typedef struct fib_destination_set_entry_t {
+    uint8_t dest[16];    /**< The destination address */
```
Can we replace the `16` with a macro?

BytesGalore replied:
sure, I would like to use `UNIVERSAL_ADDRESS_SIZE` for this magic number when #3022 is merged :) 
We can add a substitute now and switch it after #3022 has been merged.

OlegHahm replied:
:+1:

OlegHahm replied:
And you can directly squash the change if you like.

BytesGalore replied:
ok, done. (added substitute and squashed directly)

OlegHahm commented:
Kicked Travis. Merge at will if Travis is green.

OlegHahm commented:
Ok, failed again for the same test, so I suppose it is a real issue.

cgundogan commented:
running unittests locally for `samr21-xpro`:
```
/usr/bin/../lib/gcc/arm-none-eabi/4.8.1/../../../../arm-none-eabi/bin/ld: $RIOTHOME/tests/unittests/bin/samr21-xpro/unittests.elf section `.text' will not fit in region `rom'
/usr/bin/../lib/gcc/arm-none-eabi/4.8.1/../../../../arm-none-eabi/bin/ld: region `rom' overflowed by 26176 bytes
```

BytesGalore commented:
@cgundogan tested build locally for `tests-fib` with samr21-xpro -> successfully
will flash it to be certain

OlegHahm commented:
@BytesGalore, if I see it correctly it's the whole unittests suite that doesn't fit any more, not the fib part alone.

BytesGalore commented:
@OlegHahm confirmed, so any propositions how to handle this? It will be keep on failing for all subsequent added unittests.

OlegHahm commented:
I think there's only one choice: add it to `BOARD_INSUFFICIENT_RAM` in `tests/unittests/Makefile`.

cgundogan commented:
@OlegHahm isn't it the rom section, which is overfowing?

OlegHahm commented:
Ah, yes, you're right. Which is actually weird.

authmillenon commented:
The macro `BOARD_INSUFFICIENT_RAM` refers to both ROM and RAM, since it just triggers the test to ignore any linking errors...

cgundogan commented:
Then we should opt to change this macro's name to be more verbose and semantically correct

OlegHahm commented:
Maybe we should rename it to `BOARD_INSUFFICIENT_MEMORY` then.

cgundogan commented:
>Maybe we should rename it to BOARD_INSUFFICIENT_MEMORY then.

+1

BytesGalore commented:
+1

BytesGalore commented:
I added the samr board to `BOARD_INSUFFICIENT_RAM`

cgundogan commented:
@BytesGalore please squash. Merge when travis is happy

cgundogan commented:
@BytesGalore  I think this PR also needs a rebase?

BytesGalore commented:
rebased and squashed

cgundogan commented:
@BytesGalore can you trigger travis?

BytesGalore commented:
hmm seems that travis is currently not available :(

cgundogan commented:
Why is travis not building this PR? It works for other PRs..

cgundogan commented:
@BytesGalore can you again rebase to master so that #2783 is also included in this branch? This would simplify the rebase of my branch in #3050 (:

cgundogan commented:
travis is green => GO
